azext_iot/digitaltwins/providers/connection/builders.py

```diff
@@ -0,0 +1,285 @@
+        self.adx_location = adx_cluster_meta["location"].lower().replace(" ", "")
+
+        adx_database_op = self.cli.invoke(
+            "rest --method get --url {}/databases/{}?{}".format(
+                self.adx_resource_id,
+                adx_database_name,
+                api_version
+            )
+        )
+        if not adx_database_op.success():
+            raise CLIError("{} retrieve Database.".format(ERROR_PREFIX))
+
+        logger.debug(FINISHED_CHECK_RESOURCE_LOG_MSG.format("Azure Data Explorer"))
+
+        self.add_dt_role_assignment(
+            role="Contributor",
+            resource_id=f"{self.adx_resource_id}/databases/{adx_database_name}"
+        )
+        self.add_adx_principal(adx_database_name, api_version)
+
+    def add_dt_role_assignment(self, role, resource_id):
+        role_str = RBAC_ROLE_MSG.format(role, resource_id)
+        logger.debug(TRY_ADD_ROLE_LOG_MSG.format(role_str))
+        if not (self.yes or prompt_y_n(msg=ADD_INPUT_MSG.format(role_str), default="y")):
+            print(SKIP_MSG.format(role_str))
+            return
+
+        current_roles_op = self.cli.invoke(
+            "role assignment list --role '{}' --assignee {} --scope {}".format(
+                role,
+                self.dt.identity.principal_id,
+                resource_id
+            )
+        )
+
+        if current_roles_op.success() and len(current_roles_op.as_json()) > 1:
+            logger.debug(PRESENT_ADD_ROLE_LOG_MSG.format(role_str))
+            return
+
+        role_command = (
+            "role assignment create --role '{}' --assignee-object-id {} "
+            "--assignee-principal-type ServicePrincipal --scope {}".format(
+                role,
+                self.dt.identity.principal_id,
+                resource_id
+            )
+        )
+        role_op = self.cli.invoke(role_command)
+        if not role_op.success():
+            print(FAIL_RBAC_MSG.format(role_str, role_command))
+            if not prompt_y_n(msg=CONT_INPUT_MSG, default="n"):
+                raise CLIError(ABORT_MSG)
+
+        logger.debug(FINISHED_ADD_ROLE_LOG_MSG.format(role_str))
+
+    def add_adx_principal(self, adx_database_name: str, api_version: str):
+        role_str = ADX_ROLE_MSG.format(adx_database_name)
+        logger.debug(TRY_ADD_ROLE_LOG_MSG.format(role_str))
+        try:
+            if not (self.yes or prompt_y_n(msg=ADD_INPUT_MSG.format(role_str), default="y")):
+                print(SKIP_MSG.format(role_str))
+                return
+
+            database_admin_list_op = self.cli.invoke(
+                "rest --method POST --url {}/databases/{}/listPrincipals?{}".format(
+                    self.adx_resource_id,
+                    adx_database_name,
+                    api_version,
+                )
+            )
+            if not database_admin_list_op.success():
+                raise database_admin_list_op.az_cli.result.error
```

Comment:
Is this raising SystemExit or are there more KPIs than that?

Reply:
This was a fun exploration I did to figure out where the error was in an embedded cli call since it wraps up the error.